<a href="https://colab.research.google.com/github/ehsanpour1/IBM/blob/master/Segmenting_and_Clustering_Neighborhoods_in_Toronto.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Segmenting and Clustering Neighborhoods in Toronto

## IBM Data Science Specialztion - Coursera

For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.

Start by creating a new Notebook for this assignment.
Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:

![alt text](https://d3c33hcgiwev3.cloudfront.net/imageAssetProxy.v1/7JXaz3NNEeiMwApe4i-fLg_40e690ae0e927abda2d4bde7d94ed133_Screen-Shot-2018-06-18-at-7.17.57-PM.png?expiry=1580860800000&hmac=8I_9VOh5baBV57Iid6cHFKiRDJXgtX_Vri_d2Ttigek)

3. To create the above dataframe:

The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.
4. Submit a link to your Notebook on your Github repository. (10 marks)

Note: There are different website scraping libraries and packages in Python. For scraping the above table, you can simply use pandas to read the table into a pandas dataframe.

Another way, which would help to learn for more complicated cases of web scraping is using the BeautifulSoup package. Here is the package's main documentation page: http://beautiful-soup-4.readthedocs.io/en/latest/

The package is so popular that there is a plethora of tutorials and examples on how to use it. Here is a very good Youtube video on how to use the BeautifulSoup package: https://www.youtube.com/watch?v=ng2o98k983k

Use pandas, or the BeautifulSoup package, or any other way you are comfortable with to transform the data in the table on the Wikipedia page into the above pandas dataframe.

In [0]:
!pip install beautifulsoup4

!pip install geopy

!pip install folium


In [0]:
import numpy as np
import pandas as pd
import requests
from pandas.io.json import json_normalize
import json

import requests

from bs4 import BeautifulSoup

from geopy.geocoders import Nominatim

import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

print('Libraries imported!')

Libraries imported!


# Part 1: Web scraping for Toronto neighborhood and build a clean dataframe

Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and 
    to transform the data into a pandas dataframe.

* The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
* Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
* More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma.
* If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
* Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
* In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

**Note:** There are different website scraping libraries and packages in Python. One of the most common packages is BeautifulSoup. Here is the package's main documentation page: http://beautiful-soup-4.readthedocs.io/en/latest/

In [0]:
# open Wiki page with Beautiful Soup
data = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(data, 'html.parser')


**Scrape the HTML**

The data we want is in a table, with 3 columns **PostalCode, Borough and Neighborhood.**

The table contains a list of postal codes in Canada where the first letter is M. Postal codes beginning with M are located within the city of Toronto in the province of Ontario.
Only the first three characters are listed, corresponding to the Forward Sortation Area.

In [37]:
postalCodeList = []
boroughList = []
neighborhoodList = []

for row in soup.find('table',{'class':'wikitable sortable'}).find_all('tr'):
    cells = row.find_all('td')
    if(len(cells) > 0):
        postalCodeList.append(cells[0].text)
        boroughList.append(cells[1].text)
        neighborhoodList.append(cells[2].text.rstrip('\n')) # remove the new line char from neighborhood cell

toronto_df=pd.DataFrame()
toronto_df['PostalCode']= postalCodeList
toronto_df['Borough']= boroughList
toronto_df['Neighborhood']= neighborhoodList
toronto_df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [38]:
#toronto_df_dropna = toronto_df[toronto_df.Borough != 'Not assigned'].reset_index(drop=True)
#toronto_df_dropna.head()

toronto_df['Borough'].replace('Not assigned', np.nan, inplace=True)
toronto_df.dropna(subset=['Borough'], inplace=True)
toronto_df_grouped = toronto_df.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(', '.join).reset_index()
toronto_df_grouped.columns = ['PostalCode', 'Borough', 'Neighborhood']
toronto_df.head()

PostalCode           Borough      Neighborhood
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront
5        M6A        North York  Lawrence Heights
6        M6A        North York    Lawrence Manor

In [39]:
na_neigh_rows = toronto_df_grouped.Neighborhood == 'Not assigned'

toronto_df_grouped.loc[na_neigh_rows, 'Neighborhood'] = toronto_df_grouped.loc[na_neigh_rows, 'Borough']
toronto_df_grouped[na_neigh_rows]

PostalCode       Borough  Neighborhood
93        M9A  Queen's Park  Queen's Park

In [40]:
toronto_df_cleaned = toronto_df_grouped
toronto_df_cleaned.shape

(103, 3)

# Part 2: Getting coordinates and add to the Toronto DataFrame

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In an older version of this course, we were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates.

geolocator = Nominatim(user_agent="tl-toronto-neigh")

postalList = toronto_df_cleaned['PostalCode'].values latList = [] longList = []

for post in postalList: location = None while(location is None): location = geolocator.geocode('{}, Toronto, Ontario'.format(post)) if(location != None): lat = location.latitude long = location.longitude print(post, lat, long) latList.append(lat) longList.append(long)

toronto_coors = [('Postal Code', postalList), ('Latitude', latList), ('Longitude', longList)] coors = pd.DataFrame.from_items(toronto_coors)

**Note:** Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

The csv file format has:
* 3 columns: Postal Code, Latitude and Longitude
* 103 rows: corresponding to 103 postal codes in our toronto dataframe

In [34]:
!wget -q -O "toronto_coordinates.csv" http://cocl.us/Geospatial_data
print('Coordinates downloaded!')
coors = pd.read_csv('toronto_coordinates.csv')

Coordinates downloaded!


In [41]:
print(coors.shape)
coors.head()

(103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

We will merge the two dataframes.
To get the correct result independant on the order of data, we need to set indexes of two dataframes to its Postal Code columns

In [42]:
toronto_df_temp = toronto_df_cleaned.set_index('PostalCode')
coors_temp = coors.set_index('Postal Code')
toronto_df_coors = pd.concat([toronto_df_temp, coors_temp], axis=1, join='inner')

toronto_df_coors.index.name = 'PostalCode'
toronto_df_coors.reset_index(inplace=True)

print(toronto_df_coors.shape)
toronto_df_coors.head()

(103, 5)


PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

# Part 3: Explore and cluster the neighborhoods in Toronto

In [43]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tl-toronto-neigh")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.653963, -79.387207.


### Using Folium to create a Map of Toronto with Boroughs markers on top

In [44]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, long, post, borough, neigh in zip(toronto_df_coors['Latitude'], toronto_df_coors['Longitude'], toronto_df_coors['PostalCode'], toronto_df_coors['Borough'], toronto_df_coors['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

Reduce the number of Boroughs to explore To reduce the numbers of calls to FourSquare API, we will only explore boroughs that have Toronto in their names.

In [45]:
toronto_boroughs = ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']
toronto_central_df = toronto_df_coors[toronto_df_coors['Borough'].isin(toronto_boroughs)].reset_index(drop=True)
print(toronto_central_df.shape)
toronto_central_df.head()

(39, 5)


PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [46]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, long, post, borough, neigh in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

### Using FourSquare API to explore the Boroughs

In [47]:
CLIENT_ID = '5LCOWRSOOBERWFQWQOF11VSTPQQU12NQK5R4ETGVES1C3ELA' # your Foursquare ID
CLIENT_SECRET = 'B0OC10NHX3IYERVXXKK3UXSN0PPODFWRT1GU2DRD4P2UULTK' # your Foursquare Secret
VERSION = '20190330' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 5LCOWRSOOBERWFQWQOF11VSTPQQU12NQK5R4ETGVES1C3ELA
CLIENT_SECRET:B0OC10NHX3IYERVXXKK3UXSN0PPODFWRT1GU2DRD4P2UULTK


In [0]:
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [49]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['PostalCode', 'Borough', 'Neighborhood', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.shape)
venues_df.head()

(1714, 9)


PostalCode       Borough Neighborhood  BoroughLatitude  BoroughLongitude  \
0        M4E  East Toronto  The Beaches        43.676357        -79.293031   
1        M4E  East Toronto  The Beaches        43.676357        -79.293031   
2        M4E  East Toronto  The Beaches        43.676357        -79.293031   
3        M4E  East Toronto  The Beaches        43.676357        -79.293031   
4        M4E  East Toronto  The Beaches        43.676357        -79.293031   

                            VenueName  VenueLatitude  VenueLongitude  \
0                   Glen Manor Ravine      43.676821      -79.293942   
1  The Big Carrot Natural Food Market      43.678879      -79.297734   
2                 Grover Pub and Grub      43.679181      -79.297215   
3                       Upper Beaches      43.680563      -79.292869   
4                 Seaspray Restaurant      43.678888      -79.298167   

       VenueCategory  
0              Trail  
1  Health Food Store  
2                Pub  
3       Neighborhood  
4   Asian Restaurant

In [50]:
venues_df.groupby(['PostalCode', 'Borough', 'Neighborhood'])['VenueName'].count()

PostalCode  Borough           Neighborhood                                                                                              
M4E         East Toronto      The Beaches                                                                                                     5
M4K         East Toronto      The Danforth West, Riverdale                                                                                   42
M4L         East Toronto      The Beaches West, India Bazaar                                                                                 18
M4M         East Toronto      Studio District                                                                                                42
M4N         Central Toronto   Lawrence Park                                                                                                   3
M4P         Central Toronto   Davisville North                                                                                                8

In [51]:
len(venues_df['VenueCategory'].unique())

230

#### Analyze venues in each area

In [52]:
# one hot encoding
toronto_central_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_central_onehot['PostalCode'] = venues_df['PostalCode'] 
toronto_central_onehot['Borough'] = venues_df['Borough'] 
toronto_central_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_central_onehot.columns[-3:]) + list(toronto_central_onehot.columns[:-3])
toronto_central_onehot = toronto_central_onehot[fixed_columns]

print(toronto_central_onehot.shape)
toronto_central_onehot.head()

(1714, 233)


PostalCode       Borough Neighborhoods  Afghan Restaurant  Airport  \
0        M4E  East Toronto   The Beaches                  0        0   
1        M4E  East Toronto   The Beaches                  0        0   
2        M4E  East Toronto   The Beaches                  0        0   
3        M4E  East Toronto   The Beaches                  0        0   
4        M4E  East Toronto   The Beaches                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 1                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera St

#### Get the frequency of occurance of each category in an area

In [53]:
toronto_central_venues_freq = toronto_central_onehot.groupby(['PostalCode', 'Borough', 'Neighborhoods']).mean().reset_index()
print(toronto_central_venues_freq.shape)
toronto_central_venues_freq.head()


(39, 233)


PostalCode          Borough                   Neighborhoods  \
0        M4E     East Toronto                     The Beaches   
1        M4K     East Toronto    The Danforth West, Riverdale   
2        M4L     East Toronto  The Beaches West, India Bazaar   
3        M4M     East Toronto                 Studio District   
4        M4N  Central Toronto                   Lawrence Park   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.023810   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.047619   
4             0.0              0.0               0.0             0.000000   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0           0.0       0.0          0.0                  0.0               0.2   
1           0.0       0.0          0.0                  0.0               0.0   
2           0.0       0.0          0.0                  0.0               0.0   
3           0.0       0.0          0.0                  0.0               0.0   
4           0.0       0.0          0.0                  0.0               0.0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                 0.0            0.0        0.0         0.0         0.0   
1                 0.0            0.0        0.0         0.0         0.0   
2                 0.0            0.0        0.0         0.0         0.0   
3                 0.0            0.0        0.0         0.0         0.0   
4                 0.0            0.0        0.0         0.0         0.0   

     Bakery     Bank      Bar  Baseball Stadium  Basketball Stadium  Beach  \
0  0.000000  0.00000  0.00000               0.0                 0.0    0.0   
1  0.023810  0.00000  0.00000               0.0                 0.0    0.0   
2  0.000000  0.00000  0.00000               0.0                 0.0    0.0   
3  0.047619  0.02381  0.02381               0.0                 0.0    0.0   
4  0.000000  0.00000  0.00000               0.0                 0.0    0.0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0              0.0       0.0         0.0                 0.0     0.0   
1              0.0       0.0         0.0                 0.0     0.0   
2              0.0       0.0         0.0                 0.0     0.0   
3              0.0       0.0         0.0                 0.0     0.0   
4              0.0       0.0         0.0                 0.0     0.0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0            0.0   0.000000       0.0                   0.0             0.0   
1            0.0   0.047619       0.0                   0.0             0.0   
2            0.0   0.000000       0.0                   0.0             0.0   
3            0.0   0.023810       0.0                   0.0             0.0   
4            0.0   0.000000       0.0                   0.0             0.0   

    Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0  0.000000          0.00000       0.0      0.000000       0.000000  0.000000   
1  0.023810          0.02381       0.0      0.000000       0.000000  0.000000   
2  0.055556          0.00000       0.0      0.055556       0.055556  0.000000   
3  0.047619          0.00000       0.0      0.000000       0.000000  0.000000   
4  0.000000          0.00

#### Get 10 most occurance venue types in each area

In [54]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['PostalCode', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = toronto_central_venues_freq['PostalCode']
neighborhoods_venues_sorted['Borough'] = toronto_central_venues_freq['Borough']
neighborhoods_venues_sorted['Neighborhoods'] = toronto_central_venues_freq['Neighborhoods']

for ind in np.arange(toronto_central_venues_freq.shape[0]):
    row_categories = toronto_central_venues_freq.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
neighborhoods_venues_sorted

PostalCode           Borough  \
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
3         M4M      East Toronto   
25        M5S  Downtown Toronto   
34        M6P      West Toronto   
6         M4R   Central Toronto   
19        M5J  Downtown Toronto   
28        M5W  Downtown Toronto   
20        M5K  Downtown Toronto   
15        M5C  Downtown Toronto   
21        M5L  Downtown Toronto   
29        M5X  Downtown Toronto   
14        M5B  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
37        M7A  Downtown Toronto   
9         M4V   Central Toronto   
11        M4X  Downtown Toronto   
18        M5H  Downtown Toronto   
35        M6R      West Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
22        M5N   Central Toronto   
5         M4P   Central Toronto   
23        M5P   Central Toronto   
4         M4N   Central Toronto   
2         M4L      East Toronto   
10        M4W  Downtown Toronto   
31        M6H      West Toronto   
36        M6S      West Toronto   
0         M4E      East Toronto   
24        M5R   Central Toronto   
7         M4S   Central Toronto   
38        M7Y      East Toronto   
8         M4T   Central Toronto   
26        M5T  Downtown Toronto   

                                        Neighborhoods  1st Most Common Venue  \
27  CN Tower, Bathurst Quay, Island airport, Harbo...         Airport Lounge   
32                           Little Portugal, Trinity                    Bar   
33       Brockton, Exhibition Place, Parkdale Village         Breakfast Spot   
3                                     Studio District                   Café   
25                     Harbord, University of Toronto                   Café   
34                      High Park, The Junction South                   Café   
6                                  North Toronto West         Clothing Store   
19  Harbourfront East, Toronto Islands, Union Station            Coffee Shop   
28                    Stn A PO Boxes 25 The Esplanade            Coffee Shop   
20           Design Exchange, Toronto Dominion Centre            Coffee Shop   
15                                     St. James Town            Coffee Shop   
21                     Commerce Court, Victoria Hotel            Coffee Shop   
29             First Canadian Place, Underground city            Coffee Shop   
14                           Ryerson, Garden District            Coffee Shop   
16                                        Berczy Park            Coffee Shop   
17                                 Central Bay Street            Coffee Shop   
12                               Church and Wellesley            Coffee Shop   
13                                       Harbourfront            Coffee Shop   
37                                       Queen's Park            Coffee Shop   
9   Deer Park, Forest Hill SE, Rathnelly, South Hi...            Coffee Shop   
11                        Cabbagetown, St. James Town            Coffee Shop   
18                           Adelaide, King, Richmond            Coffee Shop   
35                             Parkdale, Roncesvalles              Gift Shop   
1                        The Danforth West, Riverdale       Greek Restaurant   
30                                           Christie          Grocery Store   
22                                           Roselawn           Home Service   
5                                    Davisville North                  Hotel   
23                Forest Hill North, Forest Hill West          Jewelry Store   
4                                       Lawrence Park                   Park   
2                      The Beaches West, India Bazaar                   Park   
10                                           Rosedale                   Park   
31                       Dovercourt Villag

**Clustering areas**

Use KMeans algorigthm, try to cluster the toronto central areas into 4 clusters

In [55]:
kclusters = 4

toronto_central_venues_freq_clustering = toronto_central_venues_freq.drop(['PostalCode', 'Borough', 'Neighborhoods'], 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_central_venues_freq_clustering)

toronto_central_clustered_df = toronto_central_df
toronto_central_clustered_df['Cluster'] = kmeans.labels_

toronto_central_clustered_df = toronto_central_clustered_df.join(neighborhoods_venues_sorted.drop(['Borough', 'Neighborhoods'], 1).set_index('PostalCode'), on='PostalCode')
toronto_central_clustered_df.sort_values(['Cluster'] + freqColumns, inplace=True)
toronto_central_clustered_df

PostalCode           Borough  \
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
3         M4M      East Toronto   
25        M5S  Downtown Toronto   
34        M6P      West Toronto   
6         M4R   Central Toronto   
19        M5J  Downtown Toronto   
28        M5W  Downtown Toronto   
20        M5K  Downtown Toronto   
15        M5C  Downtown Toronto   
21        M5L  Downtown Toronto   
29        M5X  Downtown Toronto   
14        M5B  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
37        M7A  Downtown Toronto   
9         M4V   Central Toronto   
11        M4X  Downtown Toronto   
18        M5H  Downtown Toronto   
35        M6R      West Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
5         M4P   Central Toronto   
23        M5P   Central Toronto   
2         M4L      East Toronto   
31        M6H      West Toronto   
36        M6S      West Toronto   
0         M4E      East Toronto   
24        M5R   Central Toronto   
7         M4S   Central Toronto   
38        M7Y      East Toronto   
26        M5T  Downtown Toronto   
22        M5N   Central Toronto   
4         M4N   Central Toronto   
10        M4W  Downtown Toronto   
8         M4T   Central Toronto   

                                         Neighborhood   Latitude  Longitude  \
27  CN Tower, Bathurst Quay, Island airport, Harbo...  43.628947 -79.394420   
32                           Little Portugal, Trinity  43.647927 -79.419750   
33       Brockton, Exhibition Place, Parkdale Village  43.636847 -79.428191   
3                                     Studio District  43.659526 -79.340923   
25                     Harbord, University of Toronto  43.662696 -79.400049   
34                      High Park, The Junction South  43.661608 -79.464763   
6                                  North Toronto West  43.715383 -79.405678   
19  Harbourfront East, Toronto Islands, Union Station  43.640816 -79.381752   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
20           Design Exchange, Toronto Dominion Centre  43.647177 -79.381576   
15                                     St. James Town  43.651494 -79.375418   
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
29             First Canadian Place, Underground city  43.648429 -79.382280   
14                           Ryerson, Garden District  43.657162 -79.378937   
16                                        Berczy Park  43.644771 -79.373306   
17                                 Central Bay Street  43.657952 -79.387383   
12                               Church and Wellesley  43.665860 -79.383160   
13                                       Harbourfront  43.654260 -79.360636   
37                                       Queen's Park  43.662301 -79.389494   
9   Deer Park, Forest Hill SE, Rathnelly, South Hi...  43.686412 -79.400049   
11                        Cabbagetown, St. James Town  43.667967 -79.367675   
18                           Adelaide, King, Richmond  43.650571 -79.384568   
35                             Parkdale, Roncesvalles  43.648960 -79.456325   
1                        The Danforth West, Riverdale  43.679557 -79.352188   
30                                           Christie  43.669542 -79.422564   
5                                    Davisville North  43.712751 -79.390197   
23                Forest Hill North, Forest Hill West  43.696948 -79.411307   
2                      The Beaches West, India Bazaar  43.668999 -79.315572   
31                       Dovercourt Village, Dufferin  43.669005 -79.442259   
36                                 Runnymede, Swansea  43.651571 -79.484450   
0                                         The Beaches  43.676357 -79.293031   
24                The Annex, North Midtown, Yorkville  43.672710 -79.40567

#### Show those clusters onto a map

In [56]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster in zip(toronto_central_clustered_df['Latitude'], toronto_central_clustered_df['Longitude'], toronto_central_clustered_df['PostalCode'], toronto_central_clustered_df['Borough'], toronto_central_clustered_df['Neighborhood'], toronto_central_clustered_df['Cluster']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters